packages/swapper/src/swappers/ButterSwap/swapperApi/getTradeRate.ts

```diff
@@ -0,0 +1,154 @@
+  const buyAssetAddress = buyAssetIsNative ? zeroAddress : buyAssetAddressRaw
+
+  const slippageTolerancePercentageDecimal = getDefaultSlippageDecimalPercentageForSwapper(
+    SwapperName.ButterSwap,
+  )
+  const slippage = bn(slippageTolerancePercentageDecimal).times(10000).toString()
+
+  const result = await getRoute(
+    fromChainId,
+    sellAssetAddress,
+    toChainId,
+    buyAssetAddress,
+    amount,
+    slippage,
+    getButterSwapAffiliate(),
+  )
+
+  if (result.isErr()) return Err(result.unwrapErr())
+  const routeResponse = result.unwrap()
+
+  if (!isRouteSuccess(routeResponse)) {
+    if (routeResponse.errno === 2003 && routeResponse.message === 'No Route Found') {
+      return Err(
+        makeSwapErrorRight({
+          message: '[getTradeRate] No route found',
+          code: butterSwapErrorToTradeQuoteError(routeResponse.errno),
+        }),
+      )
+    }
+    return Err(
+      makeSwapErrorRight({
+        message: `[getTradeRate] ${routeResponse.message}`,
+        code: butterSwapErrorToTradeQuoteError(routeResponse.errno),
+      }),
+    )
+  }
+
+  const route = routeResponse.data[0]
+  if (!route) {
+    return Err(
+      makeSwapErrorRight({
+        message: '[getTradeRate] No route found',
+        code: TradeQuoteError.NoRouteFound,
+      }),
+    )
+  }
+
+  // Determine input and output for rate calculation
+  const inputAmount = bnOrZero(route.srcChain.totalAmountIn)
+  // Prefer dstChain if present, else bridgeChain, else srcChain
+  const outputAmount =
+    route.dstChain?.totalAmountOut ??
+    route.bridgeChain?.totalAmountOut ??
+    route.srcChain.totalAmountOut
+
+  const rate = inputAmount.gt(0) ? bnOrZero(outputAmount).div(inputAmount).toString() : '0'
+
+  const feeAsset = _deps.assetsById[feeAssetId]
+  if (!feeAsset) {
+    return Err(
+      makeSwapErrorRight({
+        message: `[getTradeRate] Fee asset not found for chainId ${sellAsset.chainId}`,
+        code: TradeQuoteError.UnsupportedChain,
+      }),
+    )
+  }
+
+  // Map gasFee.amount to networkFeeCryptoBaseUnit using fee asset precision
+  const networkFeeCryptoBaseUnit = bnOrZero(route.gasFee?.amount).gt(0)
+    ? toBaseUnit(route.gasFee.amount, feeAsset.precision)
+    : '0'
+
+  // Always a single step
+  const step = {
+    rate,
+    buyAmountBeforeFeesCryptoBaseUnit: toBaseUnit(outputAmount, buyAsset.precision),
+    buyAmountAfterFeesCryptoBaseUnit: toBaseUnit(outputAmount, buyAsset.precision),
+    sellAmountIncludingProtocolFeesCryptoBaseUnit,
+    feeData: {
+      networkFeeCryptoBaseUnit,
+      protocolFees: undefined,
+    },
+    source: SwapperName.ButterSwap,
+    buyAsset,
+    sellAsset,
+    accountNumber,
+    allowanceContract: route.contract ?? '0x0',
+    estimatedExecutionTimeMs: route.timeEstimated * 1000,
+  }
+
+  const tradeRate: TradeRate = {
+    id: route.hash,
+    rate,
+    swapperName: SwapperName.ButterSwap,
+    receiveAddress,
+    affiliateBps: affiliateBps ?? '0',
```

Comment:
This is guaranteed to be defined

Reply:
https://github.com/shapeshift/web/pull/9799/commits/e84c9b2afe2186747437e9f4b302fd02a35857b0